src/linear_algebra/matrix/fpow.lean

```diff
@@ -0,0 +1,397 @@
+
+lemma zero_fpow : ∀ z : ℤ, z ≠ 0 → (0 : M) ^ z = 0
+| (n : ℕ) h := by { rw [gpow_coe_nat, zero_pow], refine lt_of_le_of_ne n.zero_le (ne.symm _),
+  simpa using h  }
+| -[1+n]  h := by simp [zero_pow n.zero_lt_succ]
+
+lemma fzero_pow_eq (n : ℤ) : (0 : M) ^ n = if n = 0 then 1 else 0 :=
+begin
+  split_ifs with h,
+  { rw [h, gpow_zero] },
+  { rw [zero_fpow _ h] }
+end
+
+theorem inv_fpow (A : M) : ∀n:ℤ, A⁻¹ ^ n = (A ^ n)⁻¹
+| (n : ℕ) := by rw [gpow_coe_nat, gpow_coe_nat, inv_pow']
+| -[1+ n] := by rw [gpow_neg_succ_of_nat, gpow_neg_succ_of_nat, inv_pow']
+
+@[simp] lemma fpow_neg_one (A : M) : A ^ (-1 : ℤ) = A⁻¹ :=
+begin
+  convert div_inv_monoid.gpow_neg' 0 A,
+  simp only [gpow_one, int.coe_nat_zero, int.coe_nat_succ, gpow_eq_pow, zero_add]
+end
+
+theorem fpow_coe_nat (A : M) (n : ℕ) : A ^ (n : ℤ) = (A ^ n) :=
+gpow_coe_nat _ _
+
+@[simp] theorem fpow_neg_coe_nat (A : M) (n : ℕ) : A ^ (-n : ℤ) = (A ^ n)⁻¹ :=
+begin
+  cases n,
+  { simp },
+  { exact div_inv_monoid.gpow_neg' _ _ }
+end
+
+
+lemma _root_.is_unit.det_fpow {A : M} (h : is_unit A.det) (n : ℤ) : is_unit (A ^ n).det :=
+begin
+  cases n,
+  { simpa using h.pow n },
+  { simpa using h.pow n.succ }
+end
+
+lemma is_unit_det_fpow_iff {A : M} {z : ℤ} :
+  is_unit (A ^ z).det ↔ is_unit A.det ∨ z = 0 :=
+begin
+  induction z using int.induction_on with z IH z IH,
+  { simp },
+  { rw [←int.coe_nat_succ, fpow_coe_nat, det_pow, is_unit_pos_pow_iff (z.zero_lt_succ),
+        ←int.coe_nat_zero, int.coe_nat_eq_coe_nat_iff],
+    simp },
+  { rw [←neg_add', ←int.coe_nat_succ, fpow_neg_coe_nat, is_unit_nonsing_inv_det_iff,
+        det_pow, is_unit_pos_pow_iff (z.zero_lt_succ), neg_eq_zero, ←int.coe_nat_zero,
+        int.coe_nat_eq_coe_nat_iff],
+    simp }
+end
+
+lemma inv_fpow' {a : M} (h : is_unit a.det) (n : ℤ) :
+  (a ⁻¹) ^ n = a ^ (-n) :=
+begin
+  induction n using int.induction_on with n IH n IH,
+  { simp },
+  { rw [←int.coe_nat_succ, fpow_coe_nat, fpow_neg_coe_nat, inv_pow'] },
+  { rw [←neg_add', ←int.coe_nat_succ, neg_neg, fpow_coe_nat, fpow_neg_coe_nat, inv_pow',
+        nonsing_inv_nonsing_inv],
+    rw det_pow,
+    exact h.pow _ }
+end
+
+theorem fpow_neg {A : M} (h : is_unit A.det) : ∀ (n : ℤ), A ^ -n = (A ^ n)⁻¹
+| (n : ℕ) := fpow_neg_coe_nat _ _
+| -[1+ n] := by { rw [gpow_neg_succ_of_nat, neg_neg_of_nat_succ, of_nat_eq_coe, fpow_coe_nat,
+                      nonsing_inv_nonsing_inv],
+                  rw det_pow,
+                  exact h.pow _ }
+
+lemma fpow_add_one {A : M} (h : is_unit A.det) : ∀ n : ℤ, A ^ (n + 1) = A ^ n * A
+| (n : ℕ)    := by simp [← int.coe_nat_succ, pow_succ']
+| -[1+0]     := by simp [int.neg_succ_of_nat_eq, fpow_neg_one, nonsing_inv_mul _ h]
+| -[1+(n+1)] := by rw [int.neg_succ_of_nat_eq, fpow_neg h, neg_add, neg_add_cancel_right,
+  fpow_neg h, ← int.coe_nat_succ, gpow_coe_nat, gpow_coe_nat, pow_succ _ (n + 1),
+  mul_eq_mul A, mul_inv_rev, mul_eq_mul, matrix.mul_assoc,
+  nonsing_inv_mul _ h, matrix.mul_one]
```

Comment:
We don't need a special case for `-1`, also we can put everything in a `calc` block for readability.
```suggestion
lemma fpow_add_one {A : M} (h : is_unit A.det) : ∀ n : ℤ, A ^ (n + 1) = A ^ n * A
| (n : ℕ)  := by simp [← int.coe_nat_succ, pow_succ']
| -((n : ℕ) + 1) :=
calc  A ^ (-(n + 1) + 1 : ℤ)
    = (A ^ n)⁻¹ : by rw [neg_add, neg_add_cancel_right, fpow_neg h, fpow_coe_nat]
... = (A ⬝ A ^ n)⁻¹ ⬝ A : by rw [mul_inv_rev, matrix.mul_assoc, nonsing_inv_mul _ h, matrix.mul_one]
... = A ^ -(n + 1 : ℤ) * A :
      by rw [fpow_neg h, ← int.coe_nat_succ, fpow_coe_nat, pow_succ, mul_eq_mul, mul_eq_mul]
```